### PR TITLE
gh-117492: Clarify documentation of `typing.Never`

### PR DESCRIPTION
**Pull request by nineteendo**
* Issue: gh-117492

----
📚 Documentation preview 📚: https://cpython-previews--117678.org.readthedocs.build/en/117678/library/typing.html#typing.Never

### CODE REVIEW

File: Doc/library/typing.rst
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
   be used for this instead. Type checkers should treat the two equivalently.
```
I think you mean "NoReturn" instead of "Never", but I'd prefer to keep this phrase out. That sort of thing can be used in a style guide, but I don't think the CPython documentation needs to be prescriptive here.

**Comment by nineteendo** [comment]
@AlexWaygood, thoughts?

File: Doc/library/typing.rst
Comment (hauntsaninja) [inline Doc/library/typing.rst]: This reads like a style suggestion; I'm not sure CPython should be suggesting style here (and if anything I'd prefer the opposite style)

Comment (nineteendo) [reply Doc/library/typing.rst]: Quoting @JelleZijlstra:

> `Never` and `NoReturn` are exactly equivalent from a type checker's perspective and we should make the docs say so clearly. I suppose if we have to give a difference, we can say that **`NoReturn` can be preferred for return types and `Never` for other places.**

And https://typing.readthedocs.io/en/latest/spec/special-types.html#never:

> The `Never` type is equivalent to `NoReturn`, which is discussed above. The **`NoReturn` type is conventionally used in return annotations of functions, and Never is typically used in other locations**, but the two types are completely interchangeable.

Either we say that, or we give an example and state `NoReturn` is kept for backwards compatibility. Otherwise, we have two similar ways to mark functions that do not return, making `NoReturn` redundant.

Comment (nineteendo) [reply Doc/library/typing.rst]: ~I think we should just deprecate `NoReturn` and remove it in 3.18.~

**Comment by nineteendo** [comment]
@JelleZijlstra, can we deprecate `NoReturn` and remove it in 3.18 (or earlier)?

**Review by AlexWaygood — DISMISSED** [review]
I'm definitely going to oppose deprecating `NoReturn`, sorry! Perhaps if we were inventing it from scratch we wouldn't give it that name, but I haven't seen anything that would justify the huge disruption for users that would be caused by a deprecation

**Comment by nineteendo** [comment]
OK, but if we keep it, we should say that `NoReturn` is preferred for return types or that it's kept for backwards compatibility and won't be removed.

**Comment by nineteendo** [comment]
Alex, I have made the requested changes; please review again, but what do we do then instead? Opinions seem to be somewhat divided.

File: Doc/library/typing.rst
Comment (nineteendo) [reply Doc/library/typing.rst]: @AlexWaygood, then what *do* we do instead?

Comment (AlexWaygood) [reply Doc/library/typing.rst]: I have heard your question and I will respond when I have a moment to sit down and think about it

**Comment by JelleZijlstra** [comment]
I'd prefer the Python docs not to make any normative claims about which one is preferred in which contexts. We should say something like the following:

- Never and NoReturn have the same meaning in the type system
- Never was added in 3.11 (or whenever we added it)
- The name "Never" emphasizes that it is a bottom type; the name "NoReturn" emphasizes the use for functions that do not return.
- Users may make a stylistic choice about which one is preferred in which context.

**Comment by AlexWaygood** [comment]
I concur with Jelle

**Comment by nineteendo** [comment]
I've removed the style suggestion, but just to clarify, this still holds right?

> `NoReturn` can also be used as a [bottom type](https://en.wikipedia.org/wiki/Bottom_type), a type that has no values. Starting in Python 3.11, the `Never` type should be used for this concept instead.

So, you shouldn't do this in 3.11 (if you're not aiming for backwards compatibility):

```python
from typing import NoReturn

def never_call_me(arg: NoReturn) -> None:
    pass
```

**Comment by nineteendo** [comment]
@JelleZijlstra

**Comment by nineteendo** [comment]
@AlexWaygood, does this look better?

**Comment by nineteendo** [comment]
Sorry for my impatience, but do we merge the documentation of `Never` & `NoReturn` or is it clear now?
I would like to finish this pull request.

**Comment by JelleZijlstra** [comment]
I'll take a look soon. I might indeed want to document the two together.

**Comment by nineteendo** [comment]
Thoughts? https://cpython-previews--117678.org.readthedocs.build/en/117678/library/typing.html#typing.Never

**Comment by wyattscarpenter** [comment]
By the way, @nineteendo thanks for taking charge of this concern and making a pull request. I might have done it myself eventually, but concluded I wasn't yet up to the task given the ambiguous nature of the subject matter. So, thanks!

**Comment by nineteendo** [comment]
I'm afraid it's going to take until 16 may until this gets reviewed (1 month before pinging the issue + 1 week waiting for a response). @JelleZijlstra seems to be quite busy.

https://devguide.python.org/getting-started/pull-request-lifecycle/#reviewing

**Comment by JelleZijlstra** [comment]
I pushed some changes:

* There is (conceptually) only a single bottom type; Never and NoReturn both represent that single conceptual type.
* Mention the return type case first, as it is by far more common in user code.
* Mention common examples of functions that would use `Never` in their annotations.
* Mention that the two are equivalent without unnecessary further verbiage.

I'll give other maintainers some time for a review.

**Comment by nineteendo** [comment]
Thanks for taking some time, this looks a lot cleaner.

**Comment by wyattscarpenter** [comment]
Thanks for your work, @JelleZijlstra; it looks great.

File: Doc/library/typing.rst
Comment (erlend-aasland) [inline Doc/library/typing.rst]: Correctly mark up using `:data:`:

```suggestion
   :data:`!Never` and :data:`!NoReturn` represent the
```

File: Doc/library/typing.rst
Comment (erlend-aasland) [inline Doc/library/typing.rst]:
```suggestion
   :data:`!Never` and :data:`!NoReturn` have the same meaning in the type system
```

File: Doc/library/typing.rst
Comment (erlend-aasland) [inline Doc/library/typing.rst]:
```suggestion
      Added :data:`NoReturn`.
```

.. or:

```suggestion
      Added :data:`!NoReturn`.
```

File: Doc/library/typing.rst
Comment (erlend-aasland) [inline Doc/library/typing.rst]:
```suggestion
      Added :data:`Never`.
```

.. or:

```suggestion
      Added :data:`!Never`.
```

**Review by hauntsaninja — APPROVED** [review]

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**
